README.md
# Twig Console Dump

[![Build Status](https://travis-ci.org/themichaelhall/twig-console-dump.svg?branch=master)](https://travis-ci.org/themichaelhall/twig-console-dump)
[![AppVeyor](https://ci.appveyor.com/api/projects/status/github/themichaelhall/twig-console-dump?branch=master&svg=true)](https://ci.appveyor.com/project/themichaelhall/twig-console-dump/branch/master)
[![codecov.io](https://codecov.io/gh/themichaelhall/twig-console-dump/coverage.svg?branch=master)](https://codecov.io/gh/themichaelhall/twig-console-dump?branch=master)
[![StyleCI](https://styleci.io/repos/154186112/shield?style=flat&branch=master)](https://styleci.io/repos/154186112)
[![License](https://poser.pugx.org/michaelhall/twig-console-dump/license)](https://packagist.org/packages/michaelhall/twig-console-dump)
[![Latest Stable Version](https://poser.pugx.org/michaelhall/twig-console-dump/v/stable)](https://packagist.org/packages/michaelhall/twig-console-dump)
[![Total Downloads](https://poser.pugx.org/michaelhall/twig-console-dump/downloads)](https://packagist.org/packages/michaelhall/twig-console-dump)

Twig extension for dumping a variable to the browser console.

## Requirements

- PHP >= 7.1
- Twig >= 2.4

## Install with composer

``` bash
$ composer require michaelhall/twig-console-dump
```

## Basic usage

After enabling this extension, the ```dump()``` function can be used to dump a variable to the browser console.

Internally, this is done via inline JavaScript using ```console.log()``` functions.

The variable is only dumped to the console if debug mode is enabled in Twig, otherwise nothing is outputted.

### Dump a variable

```
{{ dump(foo) }}
```

#### Result

<p>
<img src="https://mh.staticfiles.se/img/console-dump-1.png" width="240" height="106" alt="Dump a variable">
</p>

### Dump a variable with a label

```
{{ dump(foo, 'Bar') }}
```

#### Result

<p>
<img src="https://mh.staticfiles.se/img/console-dump-2.png" width="240" height="106" alt="Dump a variable with a label">
</p>

### Additional options

```
{{ dump(foo, 'Bar', {'script-nonce': 'baz'}) }}
```

- script-nonce: The ```nonce``` attribute to use in the inline ```<script>``` tag for CSP-protection.

## License

MIT

CHANGELOG.md
## Unreleased

## 1.0.0 - 2018-11-05
- First stable revision.


tests/TwigConsoleDumpTest.php
<?php

declare(strict_types=1);

namespace MichaelHall\TwigConsoleDump\Tests;

use MichaelHall\TwigConsoleDump\Tests\Helpers\BasicTestClass;
use MichaelHall\TwigConsoleDump\Tests\Helpers\DerivedTestClass;
use MichaelHall\TwigConsoleDump\Tests\Helpers\EmptyTestClass;
use MichaelHall\TwigConsoleDump\Tests\Helpers\Recursive1TestClass;
use MichaelHall\TwigConsoleDump\Tests\Helpers\Recursive2TestClass;
use MichaelHall\TwigConsoleDump\Tests\Helpers\StringableTestClass;
use MichaelHall\TwigConsoleDump\TwigConsoleDump;
use PHPUnit\Framework\TestCase;
use Twig\Environment;
use Twig\Loader\ArrayLoader;

/**
 * Test TwigConsoleDump extension.
 */
class TwigConsoleDumpTest extends TestCase
{
    /**
     * Test for scalars (and null).
     *
     * @dataProvider scalarsDataProvider
     *
     * @param mixed  $var            The variable to debug.
     * @param string $expectedResult The expected result.
     */
    public function testScalars($var, string $expectedResult)
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => $var]);

        self::assertSame($expectedResult, $result);
    }

    /**
     * Data provider for testScalars.
     *
     * @return array
     */
    public function scalarsDataProvider()
    {
        return [
            [null, '<script>console.log(\'%cnull\',\'color:#555;font-weight:400\');</script>'],
            [false, '<script>console.log(\'%cfalse %cbool\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');</script>'],
            [true, '<script>console.log(\'%ctrue %cbool\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');</script>'],
            [100, '<script>console.log(\'%c100 %cint\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');</script>'],
            [-0.5, '<script>console.log(\'%c-0.5 %cfloat\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');</script>'],
            ['Foo Bar Baz', '<script>console.log(\'%c\\\'Foo Bar Baz\\\' %cstring[11]\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');</script>'],
        ];
    }

    /**
     * Test output is empty in non-debug mode.
     */
    public function testOutputIsEmptyInNonDebugMode()
    {
        $this->twigEnvironment->disableDebug();
        $result = $this->twigEnvironment->render('test.twig', ['var' => 'Foo']);

        self::assertSame('', $result);
    }

    /**
     * Test logging with escaped characters.
     */
    public function testEscapedCharacters()
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => '<p> \'Foo\' "Bar" \\ New' . "\r\n" . 'Line %c']);

        self::assertSame('<script>console.log(\'%c\\\'\\<p\\> \\\'Foo\\\' "Bar" \\\\ New\\r\\nLine %%c\\\' %cstring[30]\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');</script>', $result);
    }

    /**
     * Test for an array.
     */
    public function testArray()
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => ['Foo' => 'Bar', 1 => [2, 'Baz' => false]]]);

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%carray[2]\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c\\\'Foo\\\' %c=> %c\\\'Bar\\\' %cstring[3]\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupCollapsed(\'%c1 %c=> %carray[2]\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c0 %c=> %c2 %cint\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c\\\'Baz\\\' %c=> %cfalse %cbool\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test for an object.
     */
    public function testObject()
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => new BasicTestClass()]);

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\BasicTestClass\',\'color:#555;font-weight:400\');' .
            'console.log(\'%cpublic %cpublicVar %c\\\'Foo\\\' %cstring[3]\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%cprotected %cprotectedVar %cfalse %cbool\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%cprivate %cprivateVar %c100.25 %cfloat\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupCollapsed(\'%cstatic\',\'color:#555;font-weight:400;font-style:italic\');' .
            'console.log(\'%cpublic %cpublicStaticVar %c42 %cint\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupCollapsed(\'%cprotected %cprotectedStaticVar %carray[2]\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c0 %c=> %c\\\'Bar\\\' %cstring[3]\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c1 %c=> %c\\\'Baz\\\' %cstring[3]\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            'console.log(\'%cprivate %cprivateStaticVar %cnull\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test for a object with a __toString method.
     */
    public function testStringableObject()
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => new StringableTestClass('Foo')]);

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%c\\\'Hello from Foo\\\' %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\StringableTestClass\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%cprivate %clabel %c\\\'Foo\\\' %cstring[3]\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test with a label.
     */
    public function testWithLabel()
    {
        $result = $this->twigEnvironment->render('test-label.twig', ['var' => ['Foo' => 'Bar']]);

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%cLabel %carray[1]\',\'color:#00b;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c\\\'Foo\\\' %c=> %c\\\'Bar\\\' %cstring[3]\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test for a resource.
     */
    public function testResource()
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => opendir(sys_get_temp_dir())]);

        self::assertSame('<script>console.log(\'%cresource\',\'color:#555;font-weight:400\');</script>', $result);
    }

    /**
     * Test for derived class.
     */
    public function testDerived()
    {
        $result = $this->twigEnvironment->render('test.twig', ['var' => new DerivedTestClass()]);

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%c\\\'Bar\\\' %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\DerivedTestClass\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupCollapsed(\'%cparent %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\BaseTestClass\',\'color:#555;font-weight:400;font-style:italic\',\'color:#555;font-weight:400\');' .
            'console.groupCollapsed(\'%cparent %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\AbstractBaseTestClass\',\'color:#555;font-weight:400;font-style:italic\',\'color:#555;font-weight:400\');' .
            'console.log(\'%cprivate %cfoo %c\\\'Foo\\\' %cstring[3]\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            'console.groupCollapsed(\'%cstatic\',\'color:#555;font-weight:400;font-style:italic\');' .
            'console.log(\'%cprotected %cbar %c\\\'Bar\\\' %cstring[3]\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            'console.groupEnd();' .
            'console.log(\'%cpublic %cbaz %c12345 %cint\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test with script-nonce.
     */
    public function testWithScriptNonce()
    {
        $result = $this->twigEnvironment->render('test-script-nonce.twig', ['var' => ['Foo' => 'Bar']]);

        self::assertSame(
            '<script nonce="abc">' .
            'console.groupCollapsed(\'%cLabel %carray[1]\',\'color:#00b;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c\\\'Foo\\\' %c=> %c\\\'Bar\\\' %cstring[3]\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test for a recursive object.
     */
    public function testRecursiveObject()
    {
        $recursive2 = new Recursive2TestClass();
        $recursive1 = new Recursive1TestClass($recursive2);
        $recursive2->recursive1 = $recursive1;

        $result = $this->twigEnvironment->render('test.twig', ['var' => $recursive1]);

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\Recursive1TestClass\',\'color:#555;font-weight:400\');' .
            'console.groupCollapsed(\'%cprivate %crecursive2 %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\Recursive2TestClass\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.log(\'%cpublic %crecursive1 %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\Recursive1TestClass %crecursion\',\'color:#555;font-weight:400;font-style:italic\',\'color:#00b;font-weight:400\',\'color:#555;font-weight:400\',\'color:#555;font-weight:400;font-style:italic\');' .
            'console.groupEnd();' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test with date classes (DateTime/DateTimeImmutable/DateInterval).
     */
    public function testDateClasses()
    {
        $result = $this->twigEnvironment->render('test.twig',
            [
                'var' => [
                    new \DateTime('2000-01-02 03:04:05', new \DateTimeZone('+0200')),
                    new \DateTimeImmutable('2006-07-08 09:10:11', new \DateTimeZone('-0400')),
                    new \DateInterval('P1Y2M3DT4H5M6S'),
                ],
            ]
        );

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%carray[3]\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c0 %c=> %c\\\'2000-01-02 03:04:05 +0200\\\' %cDateTime\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c1 %c=> %c\\\'2006-07-08 09:10:11 -0400\\\' %cDateTimeImmutable\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c2 %c=> %c\\\'1y 2m 3d 4h 5m 6s\\\' %cDateInterval\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#900;font-weight:600\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Test with empty content.
     */
    public function testEmptyContent()
    {
        $result = $this->twigEnvironment->render('test.twig',
            [
                'var' => [
                    [],
                    new EmptyTestClass(),
                ],
            ]
        );

        self::assertSame(
            '<script>' .
            'console.groupCollapsed(\'%carray[2]\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c0 %c=> %carray[0]\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.log(\'%c1 %c=> %cMichaelHall\\\\TwigConsoleDump\\\\Tests\\\\Helpers\\\\EmptyTestClass\',\'color:#608;font-weight:600\',\'color:#555;font-weight:400\',\'color:#555;font-weight:400\');' .
            'console.groupEnd();' .
            '</script>', $result
        );
    }

    /**
     * Set up.
     */
    public function setUp()
    {
        $arrayLoader = new ArrayLoader([
            'test.twig'              => '{{ dump(var) }}',
            'test-label.twig'        => '{{ dump(var, \'Label\') }}',
            'test-script-nonce.twig' => '{{ dump(var, \'Label\', {\'script-nonce\':\'abc\'}) }}',
        ]);
        $this->twigEnvironment = new Environment($arrayLoader, ['debug' => true]);
        $this->twigEnvironment->addExtension(new TwigConsoleDump());
    }

    /**
     * Tear down.
     */
    public function tearDown()
    {
        $this->twigEnvironment = null;
    }

    /**
     * @var Environment My Twig environment.
     */
    private $twigEnvironment;
}


tests/Helpers/BaseTestClass.php
<?php

declare(strict_types=1);

namespace MichaelHall\TwigConsoleDump\Tests\Helpers;

/**
 * A base class.
 */
class BaseTestClass extends AbstractBaseTestClass
{
    /**
     * BaseTestClass constructor.
     */
    public function __construct()
    {
        parent::__construct();

        self::$bar = 'Bar';
    }

    /**
     * @return string
     */
    public function __toString(): string
    {
        return self::$bar;
    }

    /**
     * @var string
     */
    protected static $bar;
}


tests/Helpers/BasicTestClass.php
<?php

declare(strict_types=1);

namespace MichaelHall\TwigConsoleDump\Tests\Helpers;

/**
 * A basic test class with various properties.
 */
class BasicTestClass
{
    /**
     * BasicTestClass constructor.
     */
    public function __construct()
    {
        $this->publicVar = 'Foo';
        $this->protectedVar = false;
        $this->privateVar = 100.25;

        self::$publicStaticVar = 42;
        self::$protectedStaticVar = ['Bar', 'Baz'];
        self::$privateStaticVar = null;
    }

    /**
     * @var string
     */
    public $publicVar;

    /**
     * @var bool
     */
    protected $protectedVar;

    /**
     * @var float
     */
    private $privateVar;

    /**
     * @var int
     */
    public static $publicStaticVar;

    /**
     * @var string[]
     */
    protected static $protectedStaticVar;

    /**
     * @var null
     */
    private static $privateStaticVar;
}


src/TwigConsoleDump.php
<?php
/**
 * This file is a part of the twig-console-dump package.
 *
 * Read more at https://github.com/themichaelhall/twig-console-dump
 */
declare(strict_types=1);

namespace MichaelHall\TwigConsoleDump;

use Twig\Environment;
use Twig\Extension\AbstractExtension;
use Twig\TwigFunction;

/**
 * Twig console dump extension.
 *
 * @since 1.0.0
 */
class TwigConsoleDump extends AbstractExtension
{
    /** @noinspection PhpMissingParentCallCommonInspection */

    /**
     * Returns the filters.
     *
     * @since 1.0.0
     *
     * @return TwigFunction[] The filters.
     */
    public function getFunctions(): array
    {
        return [
            new TwigFunction('dump', [$this, 'dumpFunction'], ['is_safe' => ['html'], 'needs_environment' => true]),
        ];
    }

    /**
     * The dump function.
     *
     * @since 1.0.0
     *
     * @param Environment $environment The Twig environment.
     * @param mixed       $var         The variable to dump.
     * @param string      $label       The label (optional).
     * @param array       $options     The options (optional). Valid options are: 'script-nonce'.
     *
     * @return string The result as a script printing to console.
     */
    public function dumpFunction(Environment $environment, $var, $label = '', $options = []): string
    {
        if (!$environment->isDebug()) {
            return '';
        }

        $content = [];
        if ($label !== '') {
            $content[] = [self::escapeString(strval($label)), self::STYLE_NAME];
        }

        $scriptNonce = strval($options['script-nonce'] ?? '');
        $result =
            '<script' . ($scriptNonce !== '' ? ' nonce="' . htmlentities($scriptNonce) . '"' : '') . '>' .
            self::varToLogString($var, $content, []) .
            '</script>';

        return $result;
    }

    /**
     * Converts a variable into a log string.
     *
     * @param mixed   $var             The variable.
     * @param array   $content         Optional content to insert before result.
     * @param mixed[] $previousObjects The previous processed objects.
     *
     * @return string The log string.
     */
    private static function varToLogString($var, $content, $previousObjects): string
    {
        // Null.
        if (is_null($var)) {
            $content[] = ['null', self::STYLE_TYPE];

            return self::toConsoleLog($content);
        }

        // Bool.
        if (is_bool($var)) {
            $content[] = [$var ? 'true' : 'false', self::STYLE_VALUE];
            $content[] = ['bool', self::STYLE_TYPE];

            return self::toConsoleLog($content);
        }

        // Int.
        if (is_int($var)) {
            $content[] = [$var, self::STYLE_VALUE];
            $content[] = ['int', self::STYLE_TYPE];

            return self::toConsoleLog($content);
        }

        // Float.
        if (is_float($var)) {
            $content[] = [$var, self::STYLE_VALUE];
            $content[] = ['float', self::STYLE_TYPE];

            return self::toConsoleLog($content);
        }

        // String.
        if (is_string($var)) {
            $content[] = [self::escapeAndQuoteString($var), self::STYLE_STRING_VALUE];
            $content[] = ['string[' . strlen($var) . ']', self::STYLE_TYPE];

            return self::toConsoleLog($content);
        }

        // Array.
        if (is_array($var)) {
            return self::arrayToLogString($var, $content, $previousObjects);
        }

        // Object.
        if (is_object($var)) {
            $reflectionClass = new \ReflectionClass($var);

            return self::objectToLogString($reflectionClass, $var, $content, true, $previousObjects);
        }

        // Other type.
        $content[] = [gettype($var), self::STYLE_TYPE];

        return self::toConsoleLog($content);
    }

    /**
     * Converts an array into a log string.
     *
     * @param array   $arr             The array.
     * @param array   $content         Optional content to insert before result.
     * @param mixed[] $previousObjects The previous processed objects.
     *
     * @return string The log string.
     */
    private static function arrayToLogString(array $arr, array $content, array $previousObjects): string
    {
        $size = count($arr);

        $content[] = ['array[' . $size . ']', self::STYLE_TYPE];

        $result = self::toConsoleLog($content, $size > 0);
        if ($size === 0) {
            return $result;
        }

        foreach ($arr as $key => $value) {
            $keyContent = [];
            if (is_string($key)) {
                $keyContent[] = [self::escapeAndQuoteString($key), self::STYLE_STRING_VALUE];
            } else {
                $keyContent[] = [$key, self::STYLE_VALUE];
            }

            $keyContent[] = ['=>', self::STYLE_ARROW];
            $result .= self::varToLogString($value, $keyContent, $previousObjects);
        }

        $result .= 'console.groupEnd();';

        return $result;
    }

    /**
     * Converts an object into log string.
     *
     * @param \ReflectionClass $reflectionClass   The reflection class of the object.
     * @param mixed            $obj               The object.
     * @param array            $content           Optional content to insert before result.
     * @param bool             $showDisplayString If true, show a string representation of object.
     * @param mixed[]          $previousObjects   The previous processed objects.
     *
     * @return string The log string.
     */
    private static function objectToLogString(\ReflectionClass $reflectionClass, $obj, array $content, bool $showDisplayString, array $previousObjects): string
    {
        // String representation of object.
        if ($showDisplayString) {
            $asString = self::objectToString($obj);
            if ($asString !== null) {
                $content[] = [self::escapeAndQuoteString($asString), self::STYLE_STRING_VALUE];
            }
        }

        // Class name header.
        $content[] = [self::escapeString($reflectionClass->getName()), self::STYLE_TYPE];

        // Stuck in infinite recursion loop?
        foreach ($previousObjects as $previousObject) {
            if ($previousObject === $obj) {
                $content[] = ['recursion', self::STYLE_NOTE];

                return self::toConsoleLog($content);
            }
        }

        // Does this object contain anything?
        $parentClass = $reflectionClass->getParentClass();
        /** @var \ReflectionProperty[] $reflectionProperties */
        $reflectionProperties = array_filter($reflectionClass->getProperties(), function (\ReflectionProperty $rp) use ($reflectionClass) {
            return $rp->getDeclaringClass()->getName() === $reflectionClass->getName();
        });
        $isEmpty = $parentClass === false && count($reflectionProperties) === 0;

        // Class header.
        $result = self::toConsoleLog($content, !$isEmpty);

        if ($isEmpty) {
            return $result;
        }

        // Parent class.
        if ($parentClass !== false) {
            $result .= self::objectToLogString($parentClass, $obj, [['parent', self::STYLE_NOTE]], false, $previousObjects);
        }

        // This object should not be processed again.
        $previousObjects[] = $obj;

        // Properties.
        $nonStaticProperties = [];
        $staticProperties = [];
        foreach ($reflectionProperties as $reflectionProperty) {
            if ($reflectionProperty->isStatic()) {
                $staticProperties[] = $reflectionProperty;
            } else {
                $nonStaticProperties[] = $reflectionProperty;
            }
        }

        $result .= self::objectPropertiesToLogString($nonStaticProperties, $obj, $previousObjects);

        if (count($staticProperties) !== 0) {
            $result .= self::toConsoleLog([['static', self::STYLE_NOTE]], true);
            $result .= self::objectPropertiesToLogString($staticProperties, $obj, $previousObjects);
            $result .= 'console.groupEnd();';
        }

        $result .= 'console.groupEnd();';

        return $result;
    }

    /**
     * Converts object properties into log string.
     *
     * @param \ReflectionProperty[] $reflectionProperties The reflection properties.
     * @param mixed                 $obj                  The object.
     * @param mixed[]               $previousObjects      The previous processed objects.
     *
     * @return string The log string.
     */
    private static function objectPropertiesToLogString(array $reflectionProperties, $obj, array $previousObjects): string
    {
        $result = '';

        foreach ($reflectionProperties as $reflectionProperty) {
            $reflectionProperty->setAccessible(true);
            $propertyContent = [];

            $visibility = '';
            if ($reflectionProperty->isPublic()) {
                $visibility = 'public';
            } elseif ($reflectionProperty->isProtected()) {
                $visibility = 'protected';
            } elseif ($reflectionProperty->isPrivate()) {
                $visibility = 'private';
            }

            $propertyContent[] = [$visibility, self::STYLE_NOTE];
            $propertyContent[] = [self::escapeString($reflectionProperty->getName()), self::STYLE_NAME];
            $result .= self::varToLogString($reflectionProperty->getValue($obj), $propertyContent, $previousObjects);
        }

        return $result;
    }

    /**
     * Creates a console log or group from log items.
     *
     * @param array $items   The log items.
     * @param bool  $isGroup If true, create a group, if false, create a log.
     *
     * @return string The console log or group.
     */
    private static function toConsoleLog(array $items, bool $isGroup = false): string
    {
        $texts = [];
        $styles = [];

        foreach ($items as $item) {
            $texts[] = '%c' . $item[0];
            $styles[] = '\'' . $item[1] . '\'';
        }

        return 'console.' . ($isGroup ? 'groupCollapsed' : 'log') . '(\'' . implode(' ', $texts) . '\',' . implode(',', $styles) . ');';
    }

    /**
     * Returns an object as a string or null if object could not be converted to a string.
     *
     * @param mixed $obj The object.
     *
     * @return null|string The object as a string or null.
     */
    private static function objectToString($obj): ?string
    {
        if (method_exists($obj, '__toString')) {
            return $obj->__toString();
        }

        if ($obj instanceof \DateTimeInterface) {
            return $obj->format('Y-m-d H:i:s O');
        }

        if ($obj instanceof \DateInterval) {
            return $obj->format('%yy %mm %dd %hh %im %ss');
        }

        return null;
    }

    /**
     * Escapes and quotes a string for console logging.
     *
     * @param string $s The original string.
     *
     * @return string The escaped and quoted string.
     */
    private static function escapeAndQuoteString(string $s): string
    {
        return '\\\'' . self::escapeString($s) . '\\\'';
    }

    /**
     * Escapes a string for console logging.
     *
     * @param string $s The original string.
     *
     * @return string The escaped string.
     */
    private static function escapeString(string $s): string
    {
        return str_replace(
            ['\\', '<', '>', '\'', "\n", "\r", '%'],
            ['\\\\', '\\<', '\\>', '\\\'', '\\n', '\\r', '%%'],
            $s
        );
    }

    /**
     * Styles for type (e.g. null, int, \Foo\Bar\Baz).
     */
    private const STYLE_TYPE = 'color:#555;font-weight:400';

    /**
     * Styles for ordinary values (e.g false, 42, 10.5).
     */
    private const STYLE_VALUE = 'color:#608;font-weight:600';

    /**
     * Style for strings (e.g. "Foo").
     */
    private const STYLE_STRING_VALUE = 'color:#900;font-weight:600';

    /**
     * Styles for arrow (e.g. =>).
     */
    private const STYLE_ARROW = 'color:#555;font-weight:400';

    /**
     * Styles for name (e.g. Model, myVar).
     */
    private const STYLE_NAME = 'color:#00b;font-weight:400';

    /**
     * Styles for note (e.g. Base class).
     */
    private const STYLE_NOTE = 'color:#555;font-weight:400;font-style:italic';
}
